### ISSUE TITLE
OpenSlide support in QuPath

### ISSUE THREAD

**Issue by petebankhead** [qupath/qupath#629]
QuPath supports reading whole slide images using two open source libraries: [Bio-Formats](https://www.openmicroscopy.org/bio-formats/) and [OpenSlide](http://openslide.org).

Of these, Bio-Formats is much more extensive and actively-maintained; OpenSlide only really supports single-plane, 8-bit RGB images and has not been meaningfully updated for a couple of years. Nevertheless, OpenSlide remains important to QuPath users because
* A few common file formats (e.g. .mrxs) are *only* supported by OpenSlide
* Where a format is supported by both Bio-Formats and OpenSlide, OpenSlide is generally much faster at reading image tiles (although may provide less metadata)
* OpenSlide offers alpha information that is relevant if regions are unscanned

The trouble is that OpenSlide is a C library, and getting it to work properly with QuPath is a challenge we haven't managed to solve... thanks to the complexities of how native dependencies - and, more importantly, their sub-dependencies - are handled on different platforms. Bug reports such as
* https://github.com/qupath/qupath/issues/355
* https://github.com/qupath/qupath/issues/628
* https://github.com/qupath/qupath/issues/627

along with breakages caused by `jpackage` updates (https://github.com/qupath/qupath/issues/615) show that the (largely manual) process initially used to try to package up the dependencies hasn't worked.

For these reasons, **we need a new and maintainable approach to include OpenSlide in QuPath in the future, or else drop support entirely**.

Alas, I'm sadly one of those developers who is quite helpless when attempting to build any native libraries with even moderate complexity... spoiled by Java and Python, I haven't been able to figure out a solution. Hence this issue and cry for help.

The biggest problem is getting a truly portable way to build OpenSlide on each platform. If we have that, we can potentially use JNI directly or JavaCPP (https://github.com/bytedeco/javacpp) / Gradle JavaCPP (https://github.com/bytedeco/gradle-javacpp). *Ideally* all dependencies would be packaged into a single .jar file and accessed either with JavaCPP / [native-lib-loader](https://github.com/scijava/native-lib-loader).

An alternative might be to try outsourcing OpenSlide support on macOS and Linux, *requiring* installation with a package manager (?) and only providing the 'official' pre-built binaries on Windows.

In considering this, it is worth asking whether we should be linking to OpenSlide only anyway.
**libvips** (https://github.com/libvips/libvips) is a much more extensive library, is actively maintained, and can also be built with OpenSlide support. Among its many features, libvips has the ability to write pyramidal images extremely fast.... which is a feature we would very much like to have, supplementing our ability to write [pyramidal OME-TIFFs using Bio-Formats](https://qupath.readthedocs.io/en/latest/docs/advanced/command_line.html).

**Comment by jcupitt** [comment]
Hello, I'm one of the libvips maintainers, I can add a couple more links.

libvips uses jhbuild to make windows and macos binaries, including openslide support. jhbuild came out of the gnome project: you describe a set of packages in xml and it'll download, patch, compile and link the library stack automatically. If you run inside a docker container, it's simple way to do things like nightly builds. 

The build scripts are here for win:

https://github.com/libvips/build-win64

And here for macos:

https://github.com/jcupitt/build-osx

There are a set of MXE (makefile-style) scripts here for win as well, some people find them simpler to work with.

https://github.com/libvips/build-win64-mxe

libvips 8.10 has added OME-TIFF pyramid support. There are some notes here:

https://libvips.github.io/libvips/2020/06/18/What's-new-in-8.10.html#pyramids

It should be quite a bit quicker than bioformats, though you may need to do some work to get the xml right.

If QuPath were to use libvips for openslide import, there would probably need to be some kind of Java->libvips link. There are various Java bindings, but I've not used them myself and I don't know how suitable they would be.

libvips is supposed to be mostly self-binding through introspection, so a full Java binding might only be a few weeks work. There's a high-quality C# binding which could be used as a model.

https://github.com/kleisauke/net-vips

It's about 3,000 lines of code.

**Comment by petebankhead** [comment]
Thank you @jcupitt! This gives a lot to explore.

**Comment by kwiechen** [comment]
In order to replace the winbuild script/cross compiling for Windows have you considered using MSYS2? Here with Precipoint/VMIC support from https://github.com/Markus-PP/openslide-vmic.git

![grafik](https://user-images.githubusercontent.com/17689755/104850920-2d48ae80-58f2-11eb-9df5-c8718577d1d2.png)

**Comment by petebankhead** [comment]
@kwiechen thanks for the suggestion, I haven't looked into this - what would it involve and what would it solve?
Currently, 'official' Windows builds are available for OpenSlide and seem to work fine. It is portability on Mac/Linux that poses the biggest problem (especially since pixman included in the last LTS Ubuntu release causes trouble).

**Comment by kwiechen** [comment]
using MSYS2 building openslide is simply calling ./configure && make &&
make install avoiding winbuild.sh/cross compilation. Did you rebuild pixman
or other openslide dependencies from source on Linux?

Am So., 17. Jan. 2021 um 18:52 Uhr schrieb Pete <notifications@github.com>:

**Comment by petebankhead** [comment]
@kwiechen As I understand it, MSYS2 is for building software on Windows. It might be handy in the future, but it won't solve the biggest problem that we need to somehow be able to create portable builds on all platforms. Currently, Windows is the least problematic because 'official' builds exist; Mac and Linux rely on package managers.

Simply rebuilding from source on Mac/Linux isn't sufficient to solve the problem either, because of the numerous sub-dependencies that must be built, and also the platform-specific way in which these are subsequently loaded on different platforms. I can specify exactly which OpenSlide shared library to load, but then this results in the sub-dependencies also being loaded (e.g. pixman, cairo, libtiff...) and it's here that system-wide versions are often picked up. This is very hard (impossible?) to control from Java alone because the loading of sub-dependencies doesn't use the Java library path.

I can get things to work on Windows/Linux/Mac fine, I just can't get things to work in a portable way. Somehow we need a streamlined process of building on *all* platforms that packages up all dependencies in a self-contained manner.

**Comment by kwiechen** [comment]
maybe the openslide-winbuild cross compiling script can be used .....

Am Mo., 18. Jan. 2021 um 10:19 Uhr schrieb Pete <notifications@github.com>:

**Comment by kwiechen** [comment]
https://github.com/microsoft/vcpkg may help to build openslide's dependencies?

**Comment by jcupitt** [comment]
macOS is pretty easy: you just build to a prefix and wrap it up with a small script and a plist file. I make a nip2 binary this way, including openslide support.

https://github.com/jcupitt/build-osx

It'll work, but users will need to tick various boxes to allow unsiogned application to run, which is annoying. A proper dev account and code signing fixes that, though I've never bothered.

For linux, a

**Comment by kwiechen** [comment]
> 
> 

> I can get things to work on Windows/Linux/Mac fine, I just can't get things to work in a portable way. Somehow we need a streamlined process of building on _all_ platforms that packages up all dependencies in a self-contained manner.


using https://spack.readthedocs.io/en/latest/package_list.html 

openslide and all its dependencies can be built with 

spack install openslide

tested with WSL/Ubuntu

**Comment by delta2golf** [comment]
The following library is under active development, but has been proven to be handy. No complex installation such as openslide. @jcupitt @petebankhead 
https://github.com/bayer-science-for-a-better-life/tiffslide

**Comment by petebankhead** [comment]
Back after a long delay...

@kwiechen I didn't know of Spack but it looks interesting. Only issue is that we don't actually need OpenSlide only, but rather [OpenSlide Java](https://github.com/openslide/openslide-java).

I created a [Homebrew tap](https://github.com/petebankhead/homebrew-qupath) that handles this, and works - but would require users to install locally and then has some awkward path issues to deal with. I found Homebrew installation on Mac straightforward... on Ubuntu, not so much.

I wonder if trying to get OpenSlide Java into some familiar package manager is the best way to go. If the installation is easy enough, I guess Linux users probably won't mind and Mac users might tolerate it (and Windows users can still have the 'official' binaries made available by the OpenSlide devs).

Alternatively, when Java's [foreign function & memory API](https://openjdk.org/jeps/424) stabilizes then perhaps it'll be possible to skip the current Java JNI stuff, and just use the existing package manager support for OpenSlide without Java.

@delta2golf tifffile is a Python library, so unfortunately won't help here since QuPath is written in Java.

@jcupitt thanks very much for these suggestions. It sounds trickier than I'd hoped, but also promising - I'll try to find a free day when my optimism levels are high enough to give this a proper go.

**Comment by bgilbert** [comment]
In OpenSlide we're working on some build system changes which may help here.  OpenSlide Java has already been converted to Meson and OpenSlide will follow soon.  Afterward I'm hoping to convert winbuild to a Meson project that can build a unified DLL with most of the dependencies statically linked, since Meson has good support for that.  That could be a good starting point for doing similar builds for Linux and macOS.

We've generally avoided putting OpenSlide Java into package managers so far.  It's not really maintained and the API probably shouldn't be considered "stable" in its current form.

I wasn't aware of JEP 424; thanks for the pointer.  I've filed https://github.com/openslide/openslide-java/issues/50 to track possibly switching to it.

**Comment by cpwardell** [comment]
I ran into an issue with metadata in qptiff format files being read in by OpenSlide vs BioFormats.  Files larger than ~1.6GB were being opened with BioFormats and appeared to be fine.  Files smaller than ~1.6GB were opened with OpenSlide and were missing magnification and pixel size data.  This was rectified by changing the settings in QuPath to force opening files using BioFormats, but I initially thought that there was something wrong with the data.

**I suggest retaining OpenSlide and allowing users to force it to be used, but never default to it?**  

Either way, I appreciate your producing this software.

**Comment by petebankhead** [comment]
Thanks @bgilbert - that sounds really promising. If I can help at all with maintaining OpenSlide Java just let me know.

@cpwardell thanks, but this is a bit off-topic & would be better [discussed this on the forum](https://forum.image.sc/tag/qupath). There are other files and formats where OpenSlide reading is more reliable and/or more performant than using Bio-Formats. That's why QuPath attempts to use sensible defaults, while also [giving you the option to override them](https://qupath.readthedocs.io/en/0.4/docs/tutorials/projects.html#add-images).

**Comment by bgilbert** [comment]
[openslide-winbuild](https://github.com/openslide/openslide-winbuild) now uses Meson to build a single unified DLL with all of its dependencies linked in.  (Well, almost all: libssp and winpthreads are generated by the compiler toolchain and remain separate DLLs for now.  One of them will go away eventually, and hopefully both.)  There's still a `build.sh` wrapper script, but it's not strictly necessary anymore; it's mostly for building the source zip and binary zips for distribution.  The substance of the build all happens in Meson, in the `meson` subdirectory.

That code might be a good starting point for macOS and/or Linux builds.  It'd surely need some adjustments, but I'd be willing to consider PRs to add support for those platforms.  Feel free to take a look; feedback would be welcome!

@petebankhead Any maintenance help with OpenSlide Java would be welcome.  I'd be happy to discuss further if you'd like.

**Comment by alanocallaghan** [comment]
We've made some progress on improving OpenSlide support, but there are a few issues that will need sorting out before release.

# Linux

- While it's easy enough to get a SO of openslide with minimal dependencies (`libpcre2-8` is a stubborn dependency)

**Comment by alanocallaghan** [comment]
We've made some progress on improving OpenSlide support, but there are a few issues that will need sorting out before release. I'll try to summarise what I've been working on in a semi-lucid way...

## Java-C interface

We plan to switch from JNI to JNA to wrap OpenSlide. This seems to reduce the complexity of loading openslide, and has minimal overhead for us, as well as probably making it easier to access new OpenSlide features.

## Build process

I have forked and modified `openslide-winbuild` to try to make binaries with fewer dynamically-linked subdependencies for Mac and Linux. I've moved some of this code to a [non-forked repo](https://github.com/alanocallaghan/openslide-crossplatform-builds), as most of it is unlikely to be useful upstream (though I'm of course happy to contribute it if so).



## Windows

The self-contained builds provided by OpenSlide work well here, as do binaries from my forked repo, with the proviso that we need to load them from a path rather than a JAR, as the JNA library loading code will not find `winpthreads` or `ssp` if we try to load from a jar on the classpath.


## Mac

x86 and M1 builds using the same meson wrap system as Windows seem to work fairly well at first glance (see [x86](https://github.com/alanocallaghan/openslide-jna/actions/runs/6175143420/job/16761349958) and [m1](https://app.circleci.com/pipelines/github/alanocallaghan/openslide-jna/26/workflows/aafe7473-6e23-4897-91bd-b66d8456e2ce/jobs/26) build test runs). However, as Pete has pointed out, they have a number of weird dependencies, some of which are homebrew directories:

```
otool -L /Users/petebankhead/Downloads/openslide-natives/openslide-natives-darwin-aarch64/darwin-aarch64/libopenslide.dylib
/Users/petebankhead/Downloads/openslide-natives/openslide-natives-darwin-aarch64/darwin-aarch64/libopenslide.dylib:
	/lib/libopenslide.0.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1319.0.0)
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 1953.255.0)
	/opt/homebrew/opt/xz/lib/liblzma.5.dylib (compatibility version 10.0.0, current version 10.4.0)
	/usr/lib/libiconv.2.dylib (compatibility version 7.0.0, current version 7.0.0)
	/opt/homebrew/opt/pcre2/lib/libpcre2-8.0.dylib (compatibility version 12.0.0, current version 12.2.0)
	/System/Library/Frameworks/Foundation.framework/Versions/C/Foundation (compatibility version 300.0.0, current version 1953.255.0)
	/System/Library/Frameworks/AppKit.framework/Versions/C/AppKit (compatibility version 45.0.0, current version 2299.30.112)
	/usr/lib/libresolv.9.dylib (compatibility version 1.0.0, current version 1.0.0)
	/System/Library/Frameworks/CoreGraphics.framework/Versions/A/CoreGraphics (compatibility version 64.0.0, current version 1690.3.3)
	/System/Library/Frameworks/CoreServices.framework/Versions/A/CoreServices (compatibility version 1.0.0, current version 1228.0.0)
	/System/Library/Frameworks/CoreText.framework/Versions/A/CoreText (compatibility version 1.0.0, current version 1.0.0)
	/usr/lib/libobjc.A.dylib (compatibility version 1.0.0, current version 228.0.0)

otool -L /Users/petebankhead/Downloads/openslide-natives/openslide-natives-darwin-x86-64/darwin-x86-64/libopenslide.dylib
/Users/petebankhead/Downloads/openslide-natives/openslide-natives-darwin-x86-64/darwin-x86-64/libopenslide.dylib:
	/lib/libopenslide.0.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/local/opt/gettext/lib/libintl.8.dylib (compatibility version 12.0.0, current version 12.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1319.0.0)
	/usr/local/opt/fontconfig/lib/libfontconfig.1.dylib (compatibility version 14.0.0, current version 14.0.0)
	/usr/local/opt/freetype/lib/libfreetype.6.dylib (compatibility version 27.0.0, current version 27.1.0)
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 1953.255.0)
	/usr/local/opt/xz/lib/liblzma.5.dylib (compatibility version 10.0.0, current version 10.4.0)
	/usr/lib/libiconv.2.dylib (compatibility version 7.0.0, current version 7.0.0)
	/usr/local/opt/pcre2/lib/libpcre2-8.0.dylib (compatibility version 12.0.0, current version 12.2.0)
	/System/Library/Frameworks/Foundation.framework/Versions/C/Foundation (compatibility version 300.0.0, current version 1953.255.0)
	/System/Library/Frameworks/AppKit.framework/Versions/C/AppKit (compatibility version 45.0.0, current version 2299.30.112)
	/usr/lib/libresolv.9.dylib (compatibility version 1.0.0, current version 1.0.0)
	/usr/local/opt/zstd/lib/libzstd.1.dylib (compatibility version 1.0.0, current version 1.5.5)
	/System/Library/Frameworks/CoreGraphics.framework/Versions/A/CoreGraphics (compatibility version 64.0.0, current version 1690.3.3)
	/System/Library/Frameworks/CoreServices.framework/Versions/A/CoreServices (compatibility version 1.0.0, current version 1228.0.0)
	/System/Library/Frameworks/CoreText.framework/Versions/A/CoreText (compatibility version 1.0.0, current version 1.0.0)
	/usr/lib/libobjc.A.dylib (compatibility version 1.0.0, current version 228.0.0)
```
and notably `/usr/local/opt/gettext/lib/libintl.8.dylib` stops him from using it on an x86-64 machine, though this doesn't seem to be a problem on CI machines.



## Linux

While it's easy enough to get a SO of openslide with minimal dependencies (`libpcre2-8` stubbornly remains dynamically linked), issues arise when trying to load this alongside JavaFX. Specifically, we get a glib warning followed by an error:
```
(java:71634): GLib-GObject-WARNING **: 14:32:40.458: cannot register existing type 'gchar'
**
GLib-GObject:ERROR:../../../meson/subprojects/glib-2.72.0/gobject/gvaluetypes.c:454:_g_value_types_init: assertion failed: (type == G_TYPE_CHAR)
```
This is less urgent, as we can just distribute QuPath as `deb`/`rpm` packages, and specify `openslide-tools` as a dependency, as we will no longer depend on `openslide-java`.


## Plan for 0.5.0 release

- Windows will use a JAR of the OpenSlide builds, which we can extract to temp files and then load.
- Linux releases will depend on `openslide-tools`; if this is installed, then `libopenslide.so` should be on the search path.
- Mac builds will require homebrew (or similar) installs of `openslide`

However, if we could resolve the issues with the mac and Linux binaries, then it'd be great to use the same (or at least similar) processes on all platforms.

Would appreciate any feedback/advice from @bgilbert and @jcupitt on the best way forward, or if yous have run into some of these issues before.

**Comment by alanocallaghan** [comment]
We did end up getting self-contained binaries for all platforms built and hosted on SciJava Maven for the 4.0.0 OpenSlide release. Hopefully with the release of the Apple M1 GitHub runners in the next few months, we can contribute these changes into the upstream build cycle.

The current status of OpenSlide in QuPath is that each platform contains a self-contained OpenSlide binary, which we hook into with JNA. This means the JNI-based OpenSlide-Java is no longer needed for us, which simplifies the build quite a bit. Users can override this by pointing QuPath to their own version of OpenSlide. These changes are available in the current v0.5.0 rc(1/2) and will be in the full 0.5.0 release of QuPath also.

Closing this as all of the necessary changes now are in different repos unrelated to QuPath core.
